### DISCUSSION TITLE
Pin assignments for ESP32 D1 Mini and mystery SX1262 + PA module

### DISCUSSION THREAD

**Discussion by imfatant** [jgromes/RadioLib#1133] [Q&A] [answered]
Here are a few pics of the gear I'm trying to set up:![SX1262b](https://github.com/jgromes/RadioLib/assets/30218879/15bc281f-c2bb-4d69-b377-ecff61b57a55)![SX1262a](https://github.com/jgromes/RadioLib/assets/30218879/9a0e5df1-7787-4b1f-b2a3-ea83382acf80)![ESP32_SX1262](https://github.com/jgromes/RadioLib/assets/30218879/03c92647-2c05-477d-871e-2bf4676a79c2)

And here's the code I'm running on this ESP32 D1 Mini (it's a modified version of `RadioLib/examples/NonArduino/ESP-IDF/main/main.cpp`):
```c++
/*
   RadioLib Non-Arduino ESP-IDF Example

   This example shows how to use RadioLib without Arduino.
   In this case, a Liligo T-BEAM (ESP32 and SX1276)
   is used.

   Can be used as a starting point to port RadioLib to any platform!
   See this API reference page for details on the RadioLib hardware abstraction
   https://jgromes.github.io/RadioLib/class_hal.html

   For full API reference, see the GitHub Pages
   https://jgromes.github.io/RadioLib/
*/
// NOTE: I have modified this file slightly

// include the library
#include <RadioLib.h>

// include the hardware abstraction layer
#include "EspHal.h"

// create a new instance of the HAL class
// (we're just configuring SPI here)
// SCK:       18
// MISO:      19
// MOSI:      23
EspHal *hal = new EspHal(18, 19, 23);  // SPI pins for my Wemos ESP32 D1 Mini + SX12662

// now we can create the radio module
// NSS pin:   18
// DIO0 pin:  26
// NRST pin:  14
// DIO1 pin:  33
// SX1276 radio = new Module(hal, 18, 26, 14, 33);

// NSS pin:     5
// BUSY pin:    33
// NRESET pin:  4
SX1262 radio = new Module(hal, 5, 33, 4, RADIOLIB_NC);  // for my ESP32 D1 Mini + SX1262. It works, kinda

static const char *TAG = "main";

// the entry point for the program
// it must be declared as "extern C" because the compiler assumes this will be a C function
extern "C" void app_main(void)
{
  // initialize just like with Arduino
  ESP_LOGI(TAG, "Initializing ... ");
  // radio.XTAL = true;  // it seems I need this for the particular SX1262 module I'm using
  // but alternatively I can pass a 0 as the 8th argument to radio.begin()
  int state = radio.begin(868.0, 500.0, 6, 5, RADIOLIB_SX126X_SYNC_WORD_PRIVATE, 22, 6, 0, false);
  if (state != RADIOLIB_ERR_NONE)
  {
    ESP_LOGI(TAG, "failed, code %d\n", state);
    while (true)
    {
      hal->delay(500);
    }
  }
  ESP_LOGI(TAG, "success!\n");

  // this message is 255 bytes
  const char* msg = "The Union Of Conscientiously Workshy Elements, a humorous critique of the relentless "
                    "glorification of hard work, celebrates intentional idleness and the joy of leisure, "
                    "advocating for a society where relaxation is esteemed, as is the art of doing nothing.";
  TickType_t wakeup = xTaskGetTickCount();
  // loop forever
  for (;;)
  {
    // send a packet
    ESP_LOGI(TAG, "Transmitting packet ... ");
    state = radio.transmit(msg);
    if (state == RADIOLIB_ERR_NONE)
    {
      // the packet was successfully transmitted
      ESP_LOGI(TAG, "success!");
    }
    else
    {
      ESP_LOGI(TAG, "failed, code %d\n", state);
    }
    // transmit every second
    xTaskDelayUntil(&wakeup, 1000 / portTICK_PERIOD_MS);
  }
}
```
It kinda works. Here's the output:
```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:7172
load:0x40078000,len:15540
load:0x40080400,len:4
0x40080400: _init at ??:?

ho 8 tail 4 room 4
load:0x40080404,len:3904
entry 0x40080640
I (31) boot: ESP-IDF v5.2.2 2nd stage bootloader
I (31) boot: compile time Jun 23 2024 06:55:09
I (31) boot: Multicore bootloader
I (35) boot: chip revision: v3.1
I (39) boot.esp32: SPI Speed      : 40MHz
I (44) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 2MB
I (53) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (62) boot: ## Label            Usage          Type ST Offset   Length
I (69) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (77) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (84) boot:  2 factory          factory app      00 00 00010000 00100000
I (92) boot: End of partition table
I (96) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=0a690h ( 42640) map
I (119) esp_image: segment 1: paddr=0001a6b8 vaddr=3ffb0000 size=02254h (  8788) load
I (122) esp_image: segment 2: paddr=0001c914 vaddr=40080000 size=03704h ( 14084) load
I (130) esp_image: segment 3: paddr=00020020 vaddr=400d0020 size=188f0h (100592) map
I (168) esp_image: segment 4: paddr=00038918 vaddr=40083704 size=092cch ( 37580) load
I (189) boot: Loaded app from partition at offset 0x10000
I (189) boot: Disabling RNG early entropy source...
I (200) cpu_start: Multicore app
I (209) cpu_start: Pro cpu start user code
I (209) cpu_start: cpu freq: 240000000 Hz
I (209) cpu_start: Application information:
I (212) cpu_start: Project name:     radio1
I (217) cpu_start: App version:      1
I (222) cpu_start: Compile time:     Jun 23 2024 06:54:35
I (228) cpu_start: ELF file SHA256:  7d35500b4...
I (233) cpu_start: ESP-IDF:          v5.2.2
I (238) cpu_start: Min chip rev:     v0.0
I (243) cpu_start: Max chip rev:     v3.99 
I (247) cpu_start: Chip rev:         v3.1
I (252) heap_init: Initializing. RAM available for dynamic allocation:
I (260) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (265) heap_init: At 3FFB2C90 len 0002D370 (180 KiB): DRAM
I (272) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (278) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (284) heap_init: At 4008C9D0 len 00013630 (77 KiB): IRAM
I (292) spi_flash: detected chip: generic
I (295) spi_flash: flash io: dio
W (299) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (313) main_task: Started on CPU0
I (323) main_task: Calling app_main()
I (323) main: Initializing ... 
I (323) gpio: GPIO[18]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (333) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (343) gpio: GPIO[23]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (343) gpio: GPIO[5]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (353) gpio: GPIO[33]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (363) gpio: GPIO[4]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (473) gpio: GPIO[4]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (2773) main: success!

I (2773) main: Transmitting packet ... 
I (3673) main: success!
I (3773) main: Transmitting packet ... 
I (4673) main: success!
I (4773) main: Transmitting packet ... 
I (5673) main: success!
I (5773) main: Transmitting packet ... 
I (6673) main: success!
```

With the pins assigned as I have them now, I can get the SX1262 transmitting the 255 byte message at a maximum rate of once per second (1000 ms). I am picking up the RF signal on my RTL-SDRv4 hooked up to SDR#, and also receiving the message (some of it, anyway), with a Heltec Lora V3 ESP32-S3 SX1262 also running RadioLib but in Arduino framework. So this works, but if I try to get it to transmit the message more frequently (by altering the call to xTaskDelayUntil()), nothing changes. It refuses to go faster.

So I think I am not assigning the pins correctly, particularly those carrying interrupt signals. The mystery SX1262 + PA module has six pins whose purposes I am uncertain about, namely:
BYP
EN
BUSY
DIO1
DIO2
DIO3

All I know so far is that connecting the BUSY pin seems to need to be passed as the `irq` argument to the radio module constructor, i.e. `Module::Module(RadioLibHal *hal, uint32_t cs, uint32_t irq, uint32_t rst, uint32_t gpio = 255UL)`. Without it, RadioLib fails to initialize the SX1262, throwing:
```
I (323) main_task: Calling app_main()
I (323) main: Initializing ... 
I (323) gpio: GPIO[18]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (333) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (343) gpio: GPIO[23]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (343) gpio: GPIO[5]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (353) gpio: GPIO[27]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (363) gpio: GPIO[4]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1433) gpio: GPIO[4]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (2443) main: failed, code -2 
```

I would be most grateful if someone could explain how these pins should be connected, and if there's anything else I need to do to ensure that the PA is enabled.

Eventually I'd like to replace the ESP32 with an ESP32-S3 while staying wholly within the ESP-IDF API, but that's for the future.

**Comment by HeadBoffin** [comment]
Wow, a LoRa setup with Nitrous AND go-faster stripes! This is over expectation of hardware with sprinkles and some handcuffs & court fines.

There will be a good technical reason why it can only do one message per second, but there's no point looking at the why because in the EU you can only send 256 bytes at DR6 once every 20 seconds and then only at 14dBm. In fact there are no jurisdictions in the world where you can use a shared ISM bands to send that frequently. At that power level & send time you will probably get a reduced fine if you just hand yourself in now, because you will be very easy to find with a directional antenna surrounded by all the other users you have enacted a denial of service on, they will be the angry ones with pitch forks.

It seems strange that you are pushing the pedal to the metal but don't understand the use of the BUSY & DIO pins. They are the carburettor in this V24, managing the flow of fuel/data. And that first picture, what the heck?

As you need to go at least a 20th the speed with a vast reduction in power, there is nothing to help with, iIn summary, this is not the technology you are looking for, neither the radio, the library or the frequency fit with shifting lots of data in the way you want.

**Comment by imfatant** [comment]
>And that first picture, what the heck?

You mean the pine furniture?

**Reply by HeadBoffin** [reply]
The blank metal with a uFl connector - didn't convey much information!

**Comment by jgromes** [comment, marked as answer]
@imfatant 

> With the pins assigned as I have them now, I can get the SX1262 transmitting the 255 byte message at a maximum rate of once per second (1000 ms). 

That can be expected, take a look at the timestamps in your log:

```
I (2773) main: Transmitting packet ... 
I (3673) main: success!
I (3773) main: Transmitting packet ... 
I (4673) main: success!
```

As you can see, it takes 900 ms just to send the packet - which is quite natural, since it is huge. I would also like to point out that since you are using 500 kHz bandwidth and transmitting as fast as you can, you have effectively built a jammer. Your local authorities may want to have a word with you about that, since now you have rendered the 868 band useless, at least in your vicinity ;)

> All I know so far is that connecting the BUSY pin seems to need to be passed as the irq argument to the radio module constructor

This is all shown in the SX126x examples.

> and if there's anything else I need to do to ensure that the PA is enabled.

Seeing as you have built a jammer and are actively trying to make it as loud as possible - no, I don't think I will.

**Comment by imfatant** [comment]
@jgromes

>This is all shown in the SX126x examples.

Ah thank you :) I should have checked these. I'd focused on the single ESP-IDF example, forgetting there are more available specifically for the SX126x. I think I've figured out BYP, EN, BUSY, DIO1 and DIO2 now.

It's lucky you pointed out that I was rather overdoing the transmission. I'll only test the gear out again when I next go spelunking, particularly the PA.
